src/Rings/AbelianClosure.jl

```diff
@@ -0,0 +1,763 @@
+end
+
+Oscar.@enable_all_show_via_expressify QabElem
+
+################################################################################
+#
+#  Singular ring
+#
+################################################################################
+
+function Oscar.singular_ring(F::QabField)
+  return Singular.CoefficientRing(F)
+end
+
+################################################################################
+#
+#  Coercion between cyclotomic fields
+#
+################################################################################
+
+function isconductor(n::Int)
+  if isodd(n)
+    return true
+  end
+  return n % 4 == 0
+end
+
+function coerce_up(K::AnticNumberField, n::Int, a::QabElem)
+  d = div(n, a.c)
+  @assert n % a.c == 0
+  #z_n^(d) = z_a
+  R = parent(parent(data(a)).pol)
+  return QabElem(evaluate(R(data(a)), gen(K)^d), n)
+end
+
+function coerce_down(K::AnticNumberField, n::Int, a::QabElem)
+  throw(Hecke.NotImplemented())
+end
+
+function make_compatible(a::QabElem, b::QabElem)
+  if a.c == b.c
+    return a,b
+  end
+  d = lcm(a.c, b.c)
+  K, = cyclotomic_field(parent(a), d)
+  return coerce_up(K, d, a), coerce_up(K, d, b)
+end
+
+################################################################################
+#
+#  Ring interface functions
+#
+################################################################################
+
+isunit(a::QabElem) = !iszero(a)
+
+################################################################################
+#
+#  Arithmetic
+#
+################################################################################
+
+function +(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return QabElem(data(a) + data(b), a.c)
+end
+
+function *(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return QabElem(data(a) * data(b), a.c)
+end
+
+function -(a::QabElem)
+  return QabElem(-data(a), a.c)
+end
+
+function ^(a::QabElem, n::Integer)
+  return QabElem(data(a)^n, a.c)
+end
+
+function ^(a::QabElem, n::fmpz)
+  return a^Int(n)
+end
+
+function -(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return QabElem(a.data-b.data, a.c)
+end
+
+function //(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return QabElem(a.data//b.data, a.c)
+end
+
+function div(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return QabElem(a.data//b.data, a.c)
+end
+
+function divexact(a::QabElem, b::QabElem)
+	a, b = make_compatible(a, b)
+  return QabElem(divexact(a.data, b.data), a.c)
+end
+
+function inv(a::QabElem)
+  return QabElem(inv(data(a)), a.c)
+end
+
+################################################################################
+#
+#  Unsafe operations
+#
+################################################################################
+
+function addeq!(c::QabElem, a::QabElem)
+	_c, _a = make_compatible(c, a)
+	addeq!(_c.data, _a.data)
+  return _c
+  #c.data = _c.data
+  #c.c = _c.c
+	#return c
+end
```

Comment:
This is where we start to see tabs instead of spaces and misaligned code.
maybe do a "\t" -> "  " on this whole file